### DISCUSSION TITLE
Can I use onnxruntime-genai with a grammar for output like llama.cpp?

### DISCUSSION THREAD

**Discussion by han-minhee** [microsoft/onnxruntime-genai#564] [Ideas]
llama.cpp can force the model to generate outputs that satisfy the given grammar format.
(which, in my understanding, is actually picking the token with the highest logit among the tokens that satisfy the grammar)
Is something like this possible with the current C or C# API?
I couldn't find the corresponding API in the [documentation](https://onnxruntime.ai/docs/genai/api/).

**Comment by baijumeswani** [comment]
Hi @han-minhee. This is something we plan to support in the future. But we do not have support for this right now.

We are currently in the process of making a plan for new features to support in the 0.4.0 release and I will add this to the pool of new features so it can be considered for 0.4.0.

**Comment by baijumeswani** [comment]
I'll move this to discussions now.

**Reply by han-minhee** [reply]
Thank you for the reply. I couldn't find the related info in the release note for 0.4.0. Is it still WIP?

**Comment by kunal-vaishnavi** [comment]
All examples have been updated in [this PR](https://github.com/microsoft/onnxruntime-genai/pull/1939). Please see the READMEs in each example for more information on grammars and tool calling.
